### PR TITLE
Randomizable constructor

### PR DESCRIPTION
**Pull request by rijobro**
Adds randomizable constructor.

### Description
This allows the parent class to store `prob` and `_do_transform`. 

I'll need this for the inverse transformation PR.

### Status
**Ready**

### CODE REVIEW

**Comment by rijobro** [comment]
@wyli any chance you could have a look at this? Hoping to get it merged so I can move onto the next set of PRs.

**Comment by wyli** [comment]
> @wyli any chance you could have a look at this? Hoping to get it merged so I can move onto the next set of PRs.

thanks I'll need more time to check the details as there's no unit test to cover the changes...

`Randomizable` is a general one, for example it's also used for shuffling the dataset https://github.com/Project-MONAI/MONAI/blob/master/monai/apps/datasets.py#L136 perhaps we could subclass it for the randomised transforms?

**Comment by rijobro** [comment]
> thanks I'll need more time to check the details as there's no unit test to cover the changes...

Sure, my metric was just that none of the previous unit tests were broken. Slightly worrying that the code format wasn't picking up any warnings with this one... https://github.com/Project-MONAI/MONAI/blob/master/monai/apps/datasets.py#L136

**Comment by rijobro** [comment]
@wyli updated to add a `RandomizableTransform` class.

File: monai/transforms/transform.py [hunk 1]
```diff
@@ -77,7 +65,6 @@ def set_random_state(
         self.R = np.random.RandomState()
         return self
 
-    @abstractmethod
     def randomize(self, data: Any) -> None:
```
Comment (wyli) [inline monai/transforms/transform.py:68]: shall we support None here and in the `RandomizableTransform`?  (if it doesn't break anything...)
```suggestion
    def randomize(self, data: Optional[Any] = None) -> None:
```

**Review by wyli — APPROVED** [review]
thanks, it looks good to me

File: monai/transforms/transform.py (hunk 1 above)
Comment (rijobro) [reply monai/transforms/transform.py:68]: I shouldn't have enabled automatic merging! I would quite like the default of `data` for like `Randomizable` and `RandomizableTransform` to default to `None`. Currently I have to use a lot of `super().randomize(None)`, but I'd rather have `super().randomize()`. However, this is incompatible with certain child classes. E.g., `RandFlip` takes `np.ndarray` as argument, which is incompatible with `None`.

The alternative is to switch all child classes from e.g., `np.ndarray` to `Optional[np.ndarray]`. What do you think?

Comment (wyli) [reply monai/transforms/transform.py:68]: thanks! it's just a very minor issue, I think the current one is also good, I forgot the mypy errors...

**Pull request merged**